src/Processors/Transforms/Streaming/WatermarkStamper.cpp

```diff
@@ -32,45 +36,73 @@ namespace
 
     auto emit = emit_query->as<ASTEmitQuery>();
     assert(emit);
 
-    if (emit->periodic_interval)
+    params.strategy = emit->watermark_strategy;
+    if (params.strategy == WatermarkStrategy::BoundedOutOfOrderness)
     {
-        if (emit->after_watermark)
-            throw Exception("Streaming doesn't support having both any watermark and periodic emit", ErrorCodes::INCORRECT_QUERY);
-
-        if (emit->delay_interval)
-            throw Exception("Streaming doesn't support having both delay and periodic emit", ErrorCodes::INCORRECT_QUERY);
+        if (!emit->delay_interval) [[unlikely]]
+            throw Exception(
+                ErrorCodes::INCORRECT_QUERY, "Missing 'DELAY {}' in emit clause after watermark strategy 'BoundedOutOfOrderness'");
 
-        if (emit->timeout_interval)
-            throw Exception("Streaming doesn't support having both timeout and periodic emit", ErrorCodes::INCORRECT_QUERY);
+        params.delay_interval = extractInterval(emit->delay_interval->as<ASTFunction>());
+    }
 
+    params.mode = WatermarkEmitMode::Unknown;
+    if (emit->periodic_interval)
+    {
         params.periodic_interval = extractInterval(emit->periodic_interval->as<ASTFunction>());
-
-        params.mode = WatermarkStamperParams::EmitMode::PERIODIC;
+        params.mode = WatermarkEmitMode::Periodic;
     }
-    else if (emit->after_watermark)
+
+    if (emit->on_update)
+        params.mode = params.mode == WatermarkEmitMode::Periodic ? WatermarkEmitMode::PeriodicOnUpdate : WatermarkEmitMode::OnUpdate;
+
+    if (emit->timeout_interval)
     {
-        if (!params.window_params)
+        if (params.mode == WatermarkEmitMode::Periodic)
             throw Exception(
-                ErrorCodes::INCORRECT_QUERY, "Watermark emit is only supported in streaming queries with streaming window function");
+                ErrorCodes::INCORRECT_QUERY,
+                "The 'TIMEOUT {}' in emit clause conflicts with watermark emit mode '{}'",
+                emit->timeout_interval->formatForErrorMessage(),
+                magic_enum::enum_name(params.mode));
 
-        params.mode = params.window_params->type == WindowType::SESSION ? WatermarkStamperParams::EmitMode::WATERMARK_PER_ROW
-                                                                        : WatermarkStamperParams::EmitMode::WATERMARK;
+        params.timeout_interval = extractInterval(emit->timeout_interval->as<ASTFunction>());
     }
-    else
-        params.mode = WatermarkStamperParams::EmitMode::NONE;
+}
 
-    if (emit->timeout_interval)
-        params.timeout_interval = extractInterval(emit->timeout_interval->as<ASTFunction>());
+void initWatermarkStrategyAndEmitModeForWindowAggr(WatermarkStrategy & strategy, WatermarkEmitMode & mode, WindowParams & params)
+{
+    switch (params.type)
+    {
+        case WindowType::TUMBLE:
+            return TumbleHelper::validateWatermarkStrategyAndEmitMode(strategy, mode, assert_cast<TumbleWindowParams &>(params));
```

Comment:
What's the benefit to use free functions compared with inheritance as we did switch and all type casts etc here